File: day14-heap-sort/heap_sort.py
from heap import *

def heap_sort(heap):
    h_swap = heap.swap
    h_cd = heap.cascade_down
    heap.lim = len(heap)
    while(heap.lim > 1):
        h_swap(0, heap.lim-1)
        heap.lim -= 1
        h_cd(0)
    return heap.leaves

File: day14-heap-sort/heapsort.c
#include <stdio.h>  // Take this out.
#include <stdlib.h>
#include "heap.h"
#include "intlist.h"

/* Done in place, then wrapped in an IntList.
 * The original heap is ruined, but memory is saved. 
 */
IntList * heap_sort(Heap *heap) 
{
	IntList *list = NULL;
	int heap_size;  /* Must keep track of original size of Heap. */
	
	if(heap != NULL) {
		heap_size = heap->ins_pos;

		while(heap->ins_pos > 1) {
			swap_elements(heap, 0, heap->ins_pos - 1);
			heap->ins_pos -= 1;
			cascade_down(heap, 0);
		}

		list = wrap_array(heap->list, heap_size);
	}
	
	return list;
}
